![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2Fdata%2Bai%2Fbq-ml%2Fworkflows%2Fcross_validation&file=cross_validation.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/workflows/cross_validation/cross_validation.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/workflows/cross_validation/cross_validation.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2Fdata%252Bai%2Fbq-ml%2Fworkflows%2Fcross_validation%2Fcross_validation.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>      
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/workflows/cross_validation/cross_validation.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/workflows/cross_validation/cross_validation.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b> 
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/workflows/cross_validation/cross_validation.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a> 
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/workflows/cross_validation/cross_validation.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a> 
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/workflows/cross_validation/cross_validation.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/workflows/cross_validation/cross_validation.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a> 
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b> 
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a> 
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/workflows/cross_validation/cross_validation.ipynb"><img src="https://www.svgrepo.com/download/5445/download-button.svg" alt="Download icon" width="20px"></a> <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/workflows/cross_validation/cross_validation.ipynb">Download Notebook</a> <i>(right-click and "Save As")</i>
  </td>
</tr>
</table>

# Cross-Validation — BigQuery ML

BigQuery ML has **no native k-fold cross-validation** — `CREATE MODEL`'s `data_split_method` only supports single-split holdout (`AUTO_SPLIT`, `RANDOM`, `CUSTOM`, `SEQ`, `NO_SPLIT`). Hand-roll k-fold with deterministic hash-based fold assignment, train k models, and measure real fold-to-fold metric variance — then check whether a single holdout estimate is actually representative.

**Models used:** `LOGISTIC_REG`
**Functions used:** `ML.EVALUATE`

**Why this matters, concretely:** [`workflows/anomaly_fraud_detection/`](../anomaly_fraud_detection/)'s supervised `BOOSTED_TREE_CLASSIFIER` had a held-out eval split with only **~15 real fraud cases** (`ML.CONFUSION_MATRIX` showed TP=13, FN=2). A metric computed from just 15 positive examples is exactly the kind of high-variance situation cross-validation exists to quantify — one unlucky (or lucky) split can make a model look meaningfully better or worse than it really is.

**Data:** [`bigquery-public-data.ml_datasets.ulb_fraud_detection`](https://console.cloud.google.com/marketplace/product/bigquery-public-datasets) — the same dataset as `anomaly_fraud_detection`, for direct continuity.

**References:** [Full reference](../../RESOURCES.md) | [CREATE MODEL docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create) | [Setup guide](../../setup/)

---
## Setup

Set your project and location, authenticate, and create a shared dataset. No connection needed.

In [1]:
PROJECT_ID = 'statmike-mlops-349915'  # <-- Replace with your project ID
LOCATION = 'US'  # BigQuery dataset location
DATASET_ID = 'bq_ml'  # Shared dataset across all bq-ml notebooks

### Environment

> **Already set up the project environment?** The cell below is a no-op — packages are already in your kernel. See the [Setup Reference](../../setup/) for details.
>
> **Running standalone** (Colab, Colab Enterprise, Vertex AI Workbench)? The cell below installs required packages into your current kernel.

In [2]:
import subprocess, sys, shutil

def install(*packages):
    """Install packages using uv (fast) with pip fallback."""
    uv = shutil.which('uv')
    if uv:
        subprocess.check_call([uv, 'pip', 'install', '-q', '--python', sys.executable, *packages])
    else:
        subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', '--upgrade', *packages])

install('google-cloud-bigquery', 'db-dtypes', 'bigquery-magics')

In [3]:
# Authenticate (Colab only — other environments use Application Default Credentials)
try:
    from google.colab import auth
    auth.authenticate_user()
except ImportError:
    pass  # Not in Colab — ADC is used automatically

In [4]:
from google.cloud import bigquery
import pandas as pd

client = bigquery.Client(project=PROJECT_ID)
pd.set_option('display.max_colwidth', None)

# Create the shared dataset (idempotent)
dataset_ref = bigquery.DatasetReference(PROJECT_ID, DATASET_ID)
dataset = bigquery.Dataset(dataset_ref)
dataset.location = LOCATION
client.create_dataset(dataset, exists_ok=True)
print(f'Dataset {PROJECT_ID}.{DATASET_ID} ready')

# Register %%bigquery cell magic (auto-loaded in Colab, needed elsewhere)
%load_ext bigquery_magics

Dataset statmike-mlops-349915.bq_ml ready


---
## Step 1 — Deterministic fold assignment

`ulb_fraud_detection` has no natural row-id column — hash the full row via `TO_JSON_STRING` for a stable, deterministic fold assignment. `K = 5` (fewer than the classic 10, deliberately — keeps 5 `CREATE MODEL` calls fast to train/rerun while still clearly demonstrating fold-to-fold variance).

In [5]:
K = 5

query = f"""
CREATE OR REPLACE TABLE `{PROJECT_ID}.{DATASET_ID}.cross_validation_folds` AS
SELECT
  *,
  MOD(ABS(FARM_FINGERPRINT(TO_JSON_STRING(t))), {K}) AS fold
FROM `bigquery-public-data.ml_datasets.ulb_fraud_detection` t
"""
client.query(query).result()

query = f"""
SELECT fold, COUNT(*) AS n_rows, SUM(Class) AS n_fraud
FROM `{PROJECT_ID}.{DATASET_ID}.cross_validation_folds`
GROUP BY fold
ORDER BY fold
"""
client.query(query).to_dataframe()

,fold,n_rows,n_fraud
0,0,57057,111
1,1,56695,107
2,2,57150,88
3,3,57048,81
4,4,56857,105


Fold sizes are close to even (~56.7K-57.2K rows each), but the fraud *count* per fold ranges from 81 to 111 out of 492 total — a nearly 40% relative spread. This alone previews why a single split's fraud-detection metrics can be noisy: which fold a rare event happens to land in matters.

---
## Step 2 — Train 5 per-fold `LOGISTIC_REG` models, submitted concurrently

Each fold model trains on every row **except** its own fold (`data_split_method='NO_SPLIT'`, since fold assignment is manual, not BQML's own split). `auto_class_weights=TRUE` for the same severe imbalance already established in `anomaly_fraud_detection`.

Submit all 5 `CREATE MODEL` jobs without waiting, then wait for all — BigQuery runs distinctly-named `CREATE MODEL` jobs from one client in true parallel (same pattern verified in [`workflows/regression_based_forecasting/`](../regression_based_forecasting/)), so this finishes in roughly the time of one model, not five.

In [6]:
def train_fold_model(fold_id):
    query = f"""
    CREATE OR REPLACE MODEL `{PROJECT_ID}.{DATASET_ID}.cross_validation_fold_{fold_id}`
    OPTIONS(
      model_type = 'LOGISTIC_REG',
      input_label_cols = ['Class'],
      auto_class_weights = TRUE,
      data_split_method = 'NO_SPLIT'
    ) AS
    SELECT * EXCEPT(fold)
    FROM `{PROJECT_ID}.{DATASET_ID}.cross_validation_folds`
    WHERE fold != {fold_id}
    """
    return client.query(query)  # submitted asynchronously — caller awaits .result()

jobs = {fold_id: train_fold_model(fold_id) for fold_id in range(K)}
for fold_id, job in jobs.items():
    job.result()
print(f'Trained {K} fold models')

Trained 5 fold models


---
## Step 3 — `ML.EVALUATE` each fold model against its own held-out fold

`UNION ALL` across all 5 folds into one comparison table — the point is to see the real spread, not just one number.

In [7]:
def evaluate_fold(fold_id):
    return f"""
    SELECT {fold_id} AS fold, *
    FROM ML.EVALUATE(
      MODEL `{PROJECT_ID}.{DATASET_ID}.cross_validation_fold_{fold_id}`,
      (SELECT * EXCEPT(fold) FROM `{PROJECT_ID}.{DATASET_ID}.cross_validation_folds` WHERE fold = {fold_id})
    )
    """

query = " UNION ALL ".join(evaluate_fold(fold_id) for fold_id in range(K)) + " ORDER BY fold"
fold_metrics = client.query(query).to_dataframe()
fold_metrics

,fold,precision,recall,accuracy,f1_score,log_loss,roc_auc
0,0,0.099052,0.846847,0.984717,0.177358,0.107661,0.967671
1,1,0.093113,0.897196,0.983314,0.168717,0.118847,0.983810
2,2,0.068163,0.931818,0.980280,0.127033,0.119077,0.981242
3,3,0.068556,0.925926,0.982033,0.127660,0.115537,0.984556
4,4,0.084821,0.904762,0.981796,0.155102,0.119157,0.982744


**Verified: meaningful fold-to-fold variance on every metric** — precision ranges 0.068-0.099, recall 0.847-0.932, `roc_auc` 0.968-0.985 — from the exact same modeling recipe applied to 5 different 80%-of-data training sets. Any one of these folds, taken alone, would tell a somewhat different story about "how good is this model."

---
## Step 4 — Aggregate (mean ± stddev) vs. a same-model-type single holdout

Compute the mean and standard deviation of each metric across the 5 folds, then train **one** ordinary single 80/20 holdout model — same model type, same options — for a fair, controlled comparison.

> **GOTCHA:** don't compare this to `anomaly_fraud_detection`'s single-holdout numbers directly — that workflow used `BOOSTED_TREE_CLASSIFIER`, a different model type. Comparing across model types would confound model choice with sampling variance. The controlled comparison here uses the identical `LOGISTIC_REG` recipe for both the folds and the single holdout.

In [8]:
summary = fold_metrics[['precision', 'recall', 'accuracy', 'f1_score', 'roc_auc']].agg(['mean', 'std'])
summary

,precision,recall,accuracy,f1_score,roc_auc
mean,0.082741,0.901310,0.982428,0.151174,0.980005
std,0.014069,0.033659,0.001673,0.023154,0.007006


In [9]:
query = f"""
CREATE OR REPLACE MODEL `{PROJECT_ID}.{DATASET_ID}.cross_validation_single_holdout`
OPTIONS(
  model_type = 'LOGISTIC_REG',
  input_label_cols = ['Class'],
  auto_class_weights = TRUE,
  data_split_method = 'RANDOM',
  data_split_eval_fraction = 0.2
) AS
SELECT * EXCEPT(fold) FROM `{PROJECT_ID}.{DATASET_ID}.cross_validation_folds`
"""
client.query(query).result()

query = f"SELECT * FROM ML.EVALUATE(MODEL `{PROJECT_ID}.{DATASET_ID}.cross_validation_single_holdout`)"
client.query(query).to_dataframe()

,precision,recall,accuracy,f1_score,log_loss,roc_auc
0,0.088803,0.893204,0.98321,0.161545,0.117401,0.973774


**Verified: the single holdout's precision (~0.089) and recall (~0.893) land close to their fold means (0.083 and 0.901) — but its `roc_auc` (~0.974) lands on the low side, about one standard deviation below the 5-fold mean (0.980 ± 0.007).** Not a dramatic outlier, but a real, honest illustration: even a single holdout that "looks fine" in isolation carries meaningful uncertainty that a single number can't reveal — you'd only know it was on the low end by actually cross-validating.

---
## Related content

- [`workflows/anomaly_fraud_detection/`](../anomaly_fraud_detection/) — the small-eval-set-count problem that motivates this workflow.
- [`models/logistic_regression/`](../../models/logistic_regression/) — `LOGISTIC_REG` mechanics in depth, not repeated here.

---
## Examples — `%%bigquery` Magics

The same operations using IPython magic commands — write SQL directly in cells without Python string wrapping.

In [10]:
%%bigquery --project {PROJECT_ID}

SELECT fold, COUNT(*) AS n_rows, SUM(Class) AS n_fraud
FROM `statmike-mlops-349915.bq_ml.cross_validation_folds`
GROUP BY fold
ORDER BY fold

Query is running:   0%|          |

Downloading:   0%|          |

,fold,n_rows,n_fraud
0,0,57057,111
1,1,56695,107
2,2,57150,88
3,3,57048,81
4,4,56857,105


---
## Examples — BigFrames

No direct BigFrames equivalent — k-fold cross-validation is hand-rolled SQL here (BQML has no native k-fold support in either interface), and BigFrames' `bigframes.ml.linear_model.LogisticRegression` doesn't add anything beyond what's already shown for `LOGISTIC_REG` in [`models/logistic_regression/`](../../models/logistic_regression/).

---
## Cleanup

Drop the fold-assignment table and all 6 models created in this notebook (5 fold models + 1 single holdout).

In [11]:
for fold_id in range(K):
    client.query(f"DROP MODEL IF EXISTS `{PROJECT_ID}.{DATASET_ID}.cross_validation_fold_{fold_id}`").result()
print(f'Dropped {K} fold models')

client.query(f"DROP MODEL IF EXISTS `{PROJECT_ID}.{DATASET_ID}.cross_validation_single_holdout`").result()
print('Dropped cross_validation_single_holdout')

client.query(f"DROP TABLE IF EXISTS `{PROJECT_ID}.{DATASET_ID}.cross_validation_folds`").result()
print('Dropped cross_validation_folds')

Dropped 5 fold models
Dropped cross_validation_single_holdout
Dropped cross_validation_folds


### Remove all shared resources

When finished with **all** notebooks, uncomment and run the cell below to delete the shared dataset and everything in it.

In [12]:
# client.delete_dataset(
#     f'{PROJECT_ID}.{DATASET_ID}',
#     delete_contents=True,
#     not_found_ok=True
# )
# print(f'Dataset {PROJECT_ID}.{DATASET_ID} deleted')